### ISSUE TITLE
can't `#include <cstdio>` with `__FILE_OFFSET_BITS=64`

### ISSUE THREAD

**Issue by DanAlbert** [android-ndk/ndk#480]
Refiling @koying's follow up on #442:

> I'm afraid this fix produced an unwanted side-effect in 15c if you `#include <cstdio>` with `__USE_FILE_OFFSET64`
```
/opt/android-dev/android-toolchain-arm-21-r15c/include/c++/4.9.x/cstdio:107:11: error: '::fgetpos' has not been declared
   using ::fgetpos;
           ^
/opt/android-dev/android-toolchain-arm-21-r15c/include/c++/4.9.x/cstdio:117:11: error: '::fsetpos' has not been declared
   using ::fsetpos;
           ^
```

**Comment by DanAlbert** [comment]
This is similar to a problem that libandroid_support solves for libc++. `<cmath>` (just as one example) tries to do this for a bunch of functions that aren't really available. `libandroid_support` patches those holes by adding declarations for anything that needs to be pulled in by `using`. There are no definitions, but it lets the header be included.

We can add all the stuff that would be hidden in the `__FILE_OFFSET_BITS=64` case to libandroid_support for libc++, but for gnustl I'm going to say just turn off `__FILE_OFFSET_BITS=64`, since gnustl doesn't use libandroid_support and will be removed from the NDK soon anyway.

**Comment by koying** [comment]
Actually, worse, as `fgetpos` and friends are not defined at all, 32 or 64 bits, if API < N

**Comment by DanAlbert** [comment]
Ugh. These hacks that should have been in libandroid_support were directly in the deprecated headers.

**Comment by DanAlbert** [comment]
Oh, no, that isn't correct. `fgetpos` was available since at least Gingerbread.

**Comment by koying** [comment]
Yeah, I know ;)
The 32bits version, at least. Down't know about the 64bits one...

**Comment by koying** [comment]
Actually, pretty sure the 64bits ones were also, as we used some hacks before unified headers to do LARGE_OFFSET

**Comment by DanAlbert** [comment]
Then I don't understand what you meant by

> Actually, worse, as `fgetpos` and friends are not defined at all, 32 or 64 bits, if API < N

**Comment by enh** [comment]
fgetpos64 wasn't available until N. _FILE_OFFSET_BITS=64 used to be a no-op until r15, so you're probably just confused.

**Comment by koying** [comment]
@DanAlbert Well, with current headers, nothing is defined.
Maybe I misread, though

```
#if defined(__USE_FILE_OFFSET64)

#if __ANDROID_API__ >= 24
int fgetpos(FILE*, fpos_t*) __RENAME(fgetpos64) __INTRODUCED_IN(24);
int fsetpos(FILE*, const fpos_t*) __RENAME(fsetpos64) __INTRODUCED_IN(24);
int fseeko(FILE*, off_t, int) __RENAME(fseeko64) __INTRODUCED_IN(24);
off_t ftello(FILE*) __RENAME(ftello64) __INTRODUCED_IN(24);
#endif /* __ANDROID_API__ >= 24 */

#  if defined(__USE_BSD)

#if __ANDROID_API__ >= 24
FILE* funopen(const void*,
              int (*)(void*, char*, int),
              int (*)(void*, const char*, int),
              fpos_t (*)(void*, fpos_t, int),
              int (*)(void*)) __RENAME(funopen64) __INTRODUCED_IN(24);
#endif /* __ANDROID_API__ >= 24 */

#  endif
#else
```

**Comment by koying** [comment]
@enh Might be that the code wasn'it, but at least it compiled (we are a multi-platform app)

**Comment by DanAlbert** [comment]
Scroll down one more line.

**Comment by koying** [comment]
Sure, but this is in the context of `__USE_FILE_OFFSET64`
Or do I really misread the #if 's ?

**Comment by enh** [comment]
i think you don't understand that until r15 you were using headers that just ignored _FILE_OFFSET_BITS. full story: https://android.googlesource.com/platform/bionic/+/master/#32_bit-ABI-bugs-is-32_bit-1

**Comment by koying** [comment]
@enh We know, trust me ;)
https://github.com/xbmc/xbmc/blob/master/tools/depends/target/samba-gplv3/samba_off64_t.patch

**Comment by DanAlbert** [comment]
> Or do I really misread the #if 's ?

Certainly one of us has :) A trivial test case that just calls `fgetpos` works fine when targeting ICS.

**Comment by koying** [comment]
With `__USE_FILE_OFFSET64` ?
I'll have to drink a beer or two and re-read, then ;)

**Comment by DanAlbert** [comment]
No, *without* `__USE_FILE_OFFSET64`, but that's precisely what I would expect. That brings us back to my original plan:

> We can add all the stuff that would be hidden in the _FILE_OFFSET_BITS=64 case to libandroid_support for libc++, but for gnustl I'm going to say just turn off _FILE_OFFSET_BITS=64, since gnustl doesn't use libandroid_support and will be removed from the NDK soon anyway.

I still encourage the beer either way :)

**Comment by koying** [comment]
So the plan is to revert to pre-r15 situation re LARGE_OFFSET, ie no support. Right?
I'm not aware of libandroid_support, but I assume it's some kind of extension library?

**Comment by DanAlbert** [comment]
If you don't need 64-bit file offsets in your app, turn off `_FILE_OFFSET_BITS=64`. If you do, keep it on. Depending on your `minSdkVersion`, you might have to work around things that weren't available in that API level *for the things you use*.

That last bit (including `<cstdio>` and friends breaks your build, even if you're not using the functions that are the problem) is the part that I think we should fix here, but only for libc++. Since `_FILE_OFFSET_BITS=64` is new to the NDK, this isn't a regression for either STL. You can get the old behavior back by just not using that define.

We'll make it work well for libc++, but gnustl hasn't been getting updates/fixes for a while now and is going to be removed in a release or two anyway.

> I'm not aware of libandroid_support, but I assume it's some kind of extension library?

Yeah, it's a library that backports the things that libc++ needs to function. We've never tested it with any of the other STLs, and don't plan to (since the others are being removed soon).

**Comment by enh** [comment]
> If you don't need 64-bit file offsets in your app, turn off _FILE_OFFSET_BITS=64

strictly: "if you don't need off_t to be 64-bit in your app, turn off _FILE_OFFSET_BITS=64"

the link he showed above was using off64_t and blah64(), corresponding to `_LARGEFILE_SOURCE`, which you don't need on Android: those functions are always available (if they existed in your target API level).

**Comment by DanAlbert** [comment]
Fixed in r16.

**Comment by dmxraj** [comment]
@DanAlbert
Does r16 Beta 1 has this fix?

Link: https://dl.google.com/android/repository/android-ndk-r16-beta1-linux-x86_64.zip

Thanks

**Comment by DanAlbert** [comment]
Yep.

**Comment by acocalypso** [comment]
problem still exists.

I'm also getting the following error as well:

E:\android-ndk\android-ndk-r16-beta1\sources\android\support\include\stdio.h
Error:(36, 25) error: invalid token at start of a preprocessor expression

It seems there are some issue with 
`#if __USE_FILE_OFFSET64 && __ANDROID_API__ < __ANDROID_API_N__`

**Comment by enh** [comment]
yeah, the fix isn't in r16beta1, but will be in r16beta2.

**Comment by DanAlbert** [comment]
My mistake, thought I did that for beta 1.

**Comment by hoxnox** [comment]
> yeah, the fix isn't in r16beta1, but will be in r16beta2.

Can anybody clarify when it will be avalilable for download?

**Comment by DanAlbert** [comment]
https://groups.google.com/forum/#!forum/android-ndk-announce

**Comment by dmxraj** [comment]
This issue is still in r16-beta2

```
D:/Android/android-ndk-r16-beta2/build//../sources/cxx-stl/gnu-libstdc++/4.9/include/cstdio:107:11: error: '::fgetpos' has not been declared
   using ::fgetpos;
           ^
D:/Android/android-ndk-r16-beta2/build//../sources/cxx-stl/gnu-libstdc++/4.9/include/cstdio:117:11: error: '::fsetpos' has not been declared
   using ::fsetpos;

```

**Comment by DanAlbert** [comment]
> We'll make it work well for libc++, but gnustl hasn't been getting updates/fixes for a while now and is going to be removed in a release or two anyway.

**Comment by deguilardi** [comment]
I am getting the same error here
Error:(137, 9) error: '::fgetpos' has not been declared

What is weird is that this only happens when generating signed APK, works fine when compiling not signed apks

**Comment by feliwir** [comment]
getting this error aswell

**Comment by DanAlbert** [comment]
With libc++? Can you provide a test case?

**Comment by deguilardi** [comment]
Just update to the latest NDK, clean and recompile the project.
The workaround I found was downgrading to NDK r15c and adding the following to the Application.mk file:
APP_DEPRECATED_HEADERS := true

This solution is not working on NDK > 15 as the following link explains:
https://android.googlesource.com/platform/ndk/+/ndk-r15-release/docs/UnifiedHeaders.md

**Comment by DanAlbert** [comment]
> recompile the project.

What project?

We have a test case here that builds fine: https://android.googlesource.com/platform/ndk/+/master/tests/build/libcxx_headers_file_offset_bits_64/jni

**Comment by botanegg** [comment]
> What project?

for example - cryptopp with `-D_FILE_OFFSET_BITS=64`
https://github.com/botanegg/androidProject

**Comment by rprichard** [comment]
> for example - cryptopp with -D_FILE_OFFSET_BITS=64
> https://github.com/botanegg/androidProject

This project is using the default STL, gnustl. I can change it to libc++:

```diff
$ git diff
diff --git a/app/build.gradle b/app/build.gradle
index d4fd40a..0efff7f 100644
--- a/app/build.gradle
+++ b/app/build.gradle
@@ -14,6 +14,7 @@ android {
         }
         externalNativeBuild {
             cmake {
+                arguments "-DANDROID_STL=c++_static"
                 cppFlags "-std=c++14 -frtti -fexceptions -D_FILE_OFFSET_BITS=64"
             }
         }
```

... but then instead of the cstdio problem, fstream breaks:

```
In file included from /home/rprichard/mess/androidProject/app/src/main/cpp/cryptopp/files.cpp:7:
In file included from /home/rprichard/mess/androidProject/app/src/main/cpp/cryptopp/files.h:16:
/home/rprichard/Android/Sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/fstream:824:9: error: use of undeclared identifier 'fseeko'
    if (fseeko(__file_, __width > 0 ? __width * __off : 0, __whence))
        ^
...
```

`fseeko` and `ftello` aren't declared because the 64-bit version isn't available on old Android versions. The gradle file declares a minSdkVersion of 15.

It's easy to reproduce this problem with a standalone toolchain:

test.cc:
```c++
#include <fstream>

void foo() {
  std::ofstream out("output.txt");
}
```

Compile with, e.g.: `~/standalone-arm-16-libc++/bin/clang++ -c test.cc -D_FILE_OFFSET_BITS=64`.

There is a `_LIBCPP_HAS_NO_OFF_T_FUNCTIONS` macro in libc++'s `fstream` that makes it use `fseek`/`ftell` instead of `fseeko`/`ftello`, and adding that macro to androidProject gets it building. That macro quietly truncates offsets to 32-bits, though, and presumably the point of using `-D_FILE_OFFSET_BITS=64` is to support big files?

**Comment by enh** [comment]
at the same time, you hear people say things like

> The workaround I found was downgrading to NDK r15c and adding the following to the Application.mk file:
> APP_DEPRECATED_HEADERS := true

showing that they don't actually care about _FILE_OFFSET_BITS=64 anyway --- because their "workaround" is just switching back to headers that ignored _FILE_OFFSET_BITS. which is just a very complicated way of not setting _FILE_OFFSET_BITS=64.

given that historical versions of the NDK ignored _FILE_OFFSET_BITS, anyone having problems when upgrading can just stop setting _FILE_OFFSET_BITS to preserve the old behavior: it's by definition not  a regression (even though on other OSes you'd expect it to be).

**Comment by xsacha** [comment]
So in r16, what is the solution to get an Android project working that attempts to use fopen64/ftello64?

I found that __USE_FILE_OFFSET64 makes it drop the 64 suffix. That is, it uses fopen and ftello,  fseeko. However Android does not seem to have ftello or fseeko.

**Comment by DanAlbert** [comment]
https://android-developers.googleblog.com/2017/09/introducing-android-native-development.html

> tl;dr: Don't set _FILE_OFFSET_BITS=64 if you want to keep the behavior present in old NDKs.

**Comment by xsacha** [comment]
I didn't have _FILE_OFFSET_BITS defined and it was trying to use fopen64/ftello64.
Defining it looks like it improved the results but I still don't have ftello/fseeko.

NDK r15c 'just works' when deprecated headers are enabled (no longer possible in r16).

**Comment by enh** [comment]
if ftello64 worked in the past, it works now. though my suspicion is that you're confused, because it wasn't introduced until API level 24:
```
off64_t ftello64(FILE* __fp) __INTRODUCED_IN(24);
```
if you mean ftello, the 32-bit off_t version is there if you don't set `_FILE_OFFSET_BITS=64`:
```
off_t ftello(FILE* __fp);
```

**Comment by xsacha** [comment]
I am targeting API 21 so it complains I don't have any of these f*64 APIs.
I also don't have any of these f*o APIs.

**Comment by DanAlbert** [comment]
> NDK r15c 'just works' when deprecated headers are enabled (no longer possible in r16).

This is exactly the case if you've set `_FILE_OFFSET_BITS=64` (or manually set `__USE_FILE_OFFSET64`). You wouldn't be the first person that either didn't know you were setting it, or had it set for them by some dependency (I think boost was doing this to some people?)

If you want to be absolutely sure, find the line that calls `ftello` and add

```c++
#if _FILE_OFFSET_BITS == 64
#error "oops, file _FILE_OFFSET_BITS == 64"
#elif defined(__USE_FILE_OFFSET64)
#error "oops, __USE_FILE_OFFSET64 is defined"
#endif
```

just above it.

**Comment by xsacha** [comment]
I can't see this defined anywhere. It only fails (fopen64/ftello64 undefined references) on linking to an executable.

**Comment by DanAlbert** [comment]
Yep, first thing minizip.c does is enable that: https://github.com/madler/zlib/blob/master/contrib/minizip/minizip.c#L18

**Comment by botanegg** [comment]
> D_FILE_OFFSET_BITS=64 is to support big files?

My real project seems compilable without D_FILE_OFFSET_BITS=64
(I build it with D_FILE_OFFSET_BITS=64 in ndk r14)
But I need to be assured that nothing will break

**Comment by DanAlbert** [comment]
> (I build it with D_FILE_OFFSET_BITS=64 in ndk r14)

As the blog post explains, this was a no-op unless you'd opted into unified headers. If you had, you most had runtime bugs rather than the build time failures you have now.

**Comment by botanegg** [comment]
OK, i found 2GB file limits in project with comment about android workaround. I will try to build w/o D_FILE_OFFSET_BITS=64

**Comment by feliwir** [comment]
Please make the NDK define "_LIBCPP_HAS_NO_OFF_T_FUNCTIONS" for older api versions. It really won't build on older API versions without it

**Comment by DanAlbert** [comment]
It does though. We have tests that show it does. If you're not seeing that, presumably your build system isn't including libandroid_support.

**Comment by feliwir** [comment]
@DanAlbert try using CMake 3.10 with Android NDK r16, libc++_static and API version 14.  As soon as i define __ANDROID_API__=14, the entire build is breaking, because of missing fseeko (and friends) in the universal headers. This got fixed by the macro i pasted above.

And no, i can't use the toolchain file provided with the NDK for good reasons.

**Comment by jaimecbernardo** [comment]
This issue seems to be back on NDK r17.
I'm testing with this `test.cpp` file:
```cc
#define _FILE_OFFSET_BITS 64
#include <cstdio>
```

I'm then building six standalone toolchains (`r15c`,`r16b`,`r17b`)x(`android-22`,`android-24`):
```sh
~/android-ndk-r15c/build/tools/make-standalone-toolchain.sh --toolchain=arm-linux-androidabi-4.9 --arch=arm --stl=libc++ --platform=android-22 --install-dir=./ndk-r15-API22
~/android-ndk-r15c/build/tools/make-standalone-toolchain.sh --toolchain=arm-linux-androidabi-4.9 --arch=arm --stl=libc++ --platform=android-24 --install-dir=./ndk-r15-API24
~/android-ndk-r16b/build/tools/make-standalone-toolchain.sh --toolchain=arm-linux-androidabi-4.9 --arch=arm --stl=libc++ --platform=android-22 --install-dir=./ndk-r16-API22
~/android-ndk-r16b/build/tools/make-standalone-toolchain.sh --toolchain=arm-linux-androidabi-4.9 --arch=arm --stl=libc++ --platform=android-24 --install-dir=./ndk-r16-API24
~/android-ndk-r17/build/tools/make-standalone-toolchain.sh --toolchain=arm-linux-androidabi-4.9 --arch=arm --stl=libc++ --platform=android-22 --install-dir=./ndk-r17-API22
~/android-ndk-r17/build/tools/make-standalone-toolchain.sh --toolchain=arm-linux-androidabi-4.9 --arch=arm --stl=libc++ --platform=android-24 --install-dir=./ndk-r17-API24
```

These toolchains are able to build `test.cpp`:
```sh
./ndk-r15-API24/bin/clang -c test.cpp
./ndk-r16-API22/bin/clang -c test.cpp
./ndk-r16-API24/bin/clang -c test.cpp
./ndk-r17-API24/bin/clang -c test.cpp
```

But these toolchains are unable to build `test.cpp`:
```
$ ./ndk-r15-API22/bin/clang -c test.cpp
In file included from test.cpp:2:
/Users/myuser/android-single-c/./ndk-r15-API22/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../include/c++/4.9.x/cstdio:137:9: error: 
      no member named 'fgetpos' in the global namespace
using ::fgetpos;
      ~~^
/Users/myuser/android-single-c/./ndk-r15-API22/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../include/c++/4.9.x/cstdio:139:9: error: 
      no member named 'fsetpos' in the global namespace
using ::fsetpos;
      ~~^
2 errors generated.

$ ./ndk-r17-API22/bin/clang -c test.cpp
In file included from test.cpp:2:
/Users/myuser/android-single-c/./ndk-r17-API22/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../include/c++/4.9.x/cstdio:135:9: error: 
      no member named 'fgetpos' in the global namespace; did you mean 'fgets'?
using ::fgetpos;
      ~~^
./ndk-r17-API22/bin/../sysroot/usr/include/stdio.h:112:7: note: 'fgets' declared
      here
char* fgets(char* __buf, int __size, FILE* __fp);
      ^
In file included from test.cpp:2:
/Users/myuser/android-single-c/./ndk-r17-API22/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../include/c++/4.9.x/cstdio:137:9: error: 
      no member named 'fsetpos' in the global namespace
using ::fsetpos;
      ~~^
2 errors generated.
```

While this issue was fixed for `r16`, it seems to have come back in `r17`.
I think this issue should be re-opened, or should a new one be created?

**Comment by DanAlbert** [comment]
This is probably deserving of an r17b :(

No timeline on that yet. We generally want to give the release a week or two of soak time (not counting last week since I/O means lots of Android developers were ooo) to make sure we catch as many regressions as possible in the hotfix rather than having to release a dozen hotfixes.

**Comment by marsbible** [comment]
waiting for r17b...

**Comment by epicstar** [comment]
I'm getting this exact problem too with boost :(

I'll patiently wait for r17b.

**Comment by DanAlbert** [comment]
> the point of libandroid_support was to implement missing standard methods not present in the lower APIs.

The point of libandroid_support is to backport just enough for libc++ to work. We'd once discussed backporting everything we could, but when we asked if that was even a thing people wanted the answer was "not if it increases APK size", which is unavoidable: https://github.com/android-ndk/ndk/issues/456.

**Comment by amartinz** [comment]
what about a support library for libc++, like it is right now and a opt-in support library for backports?

this does not increase code size for people who care about it but is really helpful for people who want to utilize features across the platforms without caring about code size.

**Comment by enh** [comment]
> what about a support library for libc++, like it is right now and a opt-in support library for backports?

i think you're misunderstanding what's going on here: no implementations have been taken away. you never had 64-bit `off_t` functions: you were just in a world where your request for them was being ignored and you were getting 32-bit functions. if that's what you _actually_ want, stop requesting the 64-bit `off_t`. folks who genuinely want 64-bit functions now get them _where available_.

the bug here is that we took away a declaration for a function you don't have, but whose declaration is required by a libc++ header. that's just a bug, and we'll fix that.

@DanAlbert --- am i supposed to be making that change or are you?

**Comment by DanAlbert** [comment]
> no implementations have been taken away. you never had 64-bit off_t functions: you were just in a world where your request for them was being ignored

To clarify: `_FILE_OFFSET_BITS=64` was a no-op because the C library didn't have support for that in the past. We added that and now the feature flag that was previously silently ignored is now in effect.

> @DanAlbert --- am i supposed to be making that change or are you?

I already have it more or less ready.

**Comment by amartinz** [comment]
> i think you're misunderstanding what's going on here

Sorry, i did not mean the 64-bit off_t, i was just speaking in generally about the whole idea of an android backport library.

Wrong place and wrong time to discuss this though :)

**Comment by smilesworld116** [comment]
Any updates on r17b?
